File: sources/iOS/ios-communications/Sources/iOSCommunications/ble/api/model/gatt/client/pmd/PmdActiveMeasurement.swift
//  Copyright © 2022 Polar. All rights reserved.

import Foundation

private let MEASUREMENT_BIT_MASK: UInt8 = 0xC0

public enum PmdActiveMeasurement: UInt8 {
    case no_measurement_active = 0
    case online_measurement_active = 1
    case offline_measurement_active = 2
    case online_offline_measurement_active = 3
    
    static func fromStatusResponse(responseByte: UInt8) -> PmdActiveMeasurement {
        let masked = responseByte & MEASUREMENT_BIT_MASK
        switch (masked >> 6) {
        case online_measurement_active.rawValue:
            return .online_measurement_active
        case offline_measurement_active.rawValue:
            return .offline_measurement_active
        case online_offline_measurement_active.rawValue:
            return .online_offline_measurement_active
        case no_measurement_active.rawValue:
            return .no_measurement_active
        default:
            return .no_measurement_active
        }
    }
}

File: sources/Android/android-communications/library/src/main/java/com/polar/androidcommunications/api/ble/model/gatt/client/pmd/PmdActiveMeasurement.kt
package com.polar.androidcommunications.api.ble.model.gatt.client.pmd

enum class PmdActiveMeasurement(private val bitMask: Int?) {
    ONLINE_MEASUREMENT_ACTIVE(bitMask = 1),
    OFFLINE_MEASUREMENT_ACTIVE(bitMask = 2),
    ONLINE_AND_OFFLINE_ACTIVE(bitMask = 3),
    NO_ACTIVE_MEASUREMENT(bitMask = 0),
    UNKNOWN_MEASUREMENT_STATUS(bitMask = null);

    companion object {
        private const val MEASUREMENT_BIT_MASK: UByte = 0xC0u

        fun fromStatusResponse(responseByte: Byte): PmdActiveMeasurement {
            val masked = responseByte.toUByte() and MEASUREMENT_BIT_MASK
            return when (masked.toInt() shr 6) {
                ONLINE_MEASUREMENT_ACTIVE.bitMask -> ONLINE_MEASUREMENT_ACTIVE
                OFFLINE_MEASUREMENT_ACTIVE.bitMask -> OFFLINE_MEASUREMENT_ACTIVE
                ONLINE_AND_OFFLINE_ACTIVE.bitMask -> ONLINE_AND_OFFLINE_ACTIVE
                NO_ACTIVE_MEASUREMENT.bitMask -> NO_ACTIVE_MEASUREMENT
                else -> UNKNOWN_MEASUREMENT_STATUS
            }
        }
    }
}